### ISSUE TITLE
Tank modifier draws excessive fluid while interacting with fluid tanks in other mods

### ISSUE THREAD

**Issue by EtSH-C2H6S** [SlimeKnights/TinkersConstruct#5257]
### Minecraft Version

1.19.2

### Forge Version

43.4.0

### Mantle Version

1.10.36

### Tinkers' Construct Version

3.8.3.39

### Describe your issue

When interacting with fluid tanks in Mekanism mod using tools with tank modifier, the fluid will be totally drained EVEN IF the tank holds more fluid than fluid capacity of the tool. Same situation also happens with fluid tanks in Industrial Foregoing.

https://github.com/user-attachments/assets/0bc1a016-6316-4e8e-a72c-c6db5ca04002

### Other mods

Mekanism -10.3.9
Just Enough Items -11.6.0.1019
TConJEI -1.2.1

### Tried reproducing with just Tinkers?

No

### Performance Enchancers

None of the above

### Searched for known issues?

Searched open issues, Searched closed issues, Checked the FAQ, Checked the in game books

**Comment by KnightMiner** [comment]
Yeah, I need more context on this one. The tools work perfectly fine with our tanks, and its the tank that implements interaction behavior on right click.

Do you also have this issue with seared tanks right clicking on the other mod's fluid tank? Also try with copper cans.

**Comment by EtSH-C2H6S** [comment]
> Yeah, I need more context on this one. The tools work perfectly fine with our tanks, and its the tank that implements interaction behavior on right click.
> 
> Do you also have this issue with seared tanks right clicking on the other mod's fluid tank? Also try with copper cans.

No, seared/scorched tank and copper cans works just fine, the issue seemed to only happen with modifiable items.
Also, this issue won't occur when interacting with fluid cell from Thermal Series.

**Comment by KnightMiner** [comment]
Does this issue happen on 1.18?

Does it happen with any mods other than Melanism? E.g. create or immersive engineering

**Comment by EtSH-C2H6S** [comment]
> Does this issue happen on 1.18?
> 
> Does it happen with any mods other than Melanism? E.g. create or immersive engineering

Now I've tested tanks from create , immersive engineering , pneumaticcraft , functional storage and also tested transfering fluid from AE2 terminal, and the issue didn't occur.It only happens with tanks in mekanism and industrial foregoing. Also this issue does happens on 1.18 (with tcon ver. 3.6.4.113).

**Comment by KnightMiner** [comment]
Okay, if the issue also existed on 1.18 that rules out the recent refactors as causing this. I am going to bet since it only happens with a small subset of mods that the issue is actually on Mekanism and Industrial Foregoing's sides, they are likely not doing enough safety checks for fluid transfer. I recommend crossposting the issue there and linking it here. If I am doing something wrong on my side, somehow it works in basically every other mod.

For some additional context for the crosspost, Tinker tools with tanks support variable amounts of fluid with a fixed max capacity, though that max capacity is determined by NBT. The tools also don't always have a tank, some NBT variants will be listed as having 0 tanks.

**Comment by KnightMiner** [comment]
Just a reminder that this issue needs a cross post if you want it resolved. Tommy knowledge there is nothing I can do on my side to fix it

**Comment by EtSH-C2H6S** [comment]
> Just a reminder that this issue needs a cross post if you want it resolved. Tommy knowledge there is nothing I can do on my side to fix it

I see , but mekanism had already stopped supporting 1.19 and close the section of sending 1.19 issue long ago.So i guess this can't be solved anymore.

**Comment by EtSH-C2H6S** [comment]
This problem still exisits in 1.20 ,so I guess i can try creating 1.20 issue? (Though mekanism lables that 1.20 verson is no longer being developed)

**Comment by KnightMiner** [comment]
Sure, key this is Mekanism knows there is something wrong and can either fix it for their future builds or explain the problem on our side.

**Comment by KnightMiner** [comment]
Crosspost was added to Mekanism: mekanism/Mekanism#8312

**Comment by EtSH-C2H6S** [comment]
> Crosspost was added to Mekanism: [mekanism/Mekanism#8312](https://github.com/mekanism/Mekanism/issues/8312)

Mekanism has fixed this issue in a dev verson of 1.20.1 . It's solved now.

**Comment by KnightMiner** [comment]
I'll leave this open for now as if the fix was doing some additional fluid stack copies, it means I should verify that I'm copying everywhere I need to.

Unfortunately the state of the fluid handler API is you have to copy as the caller and as the consumer as you never know if other callers/consumers are copying.

**Comment by thiakil** [comment]
Yeah, safest bet is we both do copies

**Comment by KnightMiner** [comment]
Looking through the code, I found one fluid stack modification. If using the execute action and the fluid you passed into `fill` was larger than the tool's capacity when the tool was initially empty, then it got shrunk down to the tool's capacity before being written to NBT. That seems like the sort of missing copy that could cause this issue assuming the value passed into `fill` was a reference to the tank's true contents.

**Comment by thiakil** [comment]
> assuming the value passed into fill was a reference to the tank's true contents.

it sure was!

**Comment by KnightMiner** [comment]
This was fixed in 3.9.2.34 for 1.20, just never left a comment it seems.

Leaving open for now as a consideration to backport.
